### PR TITLE
updated early stopping docs

### PR DESCRIPTION
**Pull request by david-alexander-white**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Documentation #1406

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/early_stopping.rst [hunk 1]
```diff
@@ -39,7 +45,7 @@ is not defined.
 .. seealso::
     :class:`~pytorch_lightning.trainer.trainer.Trainer`
 
-Disable Early Stopping
+Disable Early Stopping With Callbacks on Epoch End
 ----------------------
```
Comment (Borda) [inline docs/source/early_stopping.rst:49]:
```suggestion
Disable Early Stopping with callbacks on epoch end
--------------------------------------------------
```

File: docs/source/early_stopping.rst [hunk 2]
```diff
@@ -1,15 +1,21 @@
...
+
+Default Epoch End Callback Behavior
 ----------------
 By default early stopping will be enabled if `'val_loss'`
 is found in :meth:`~pytorch_lightning.core.lightning.LightningModule.validation_epoch_end`'s
 return dict. Otherwise training will proceed with early stopping disabled.
 
-Enable Early Stopping
+Enable Early Stopping Using Callbacks On Epoch End
 ---------------------
```
Comment (Borda) [inline docs/source/early_stopping.rst:17]:
```suggestion
Enable Early Stopping using Callbacks on epoch end
--------------------------------------------------
```

File: docs/source/early_stopping.rst [hunk 3]
```diff
@@ -1,15 +1,21 @@
...
 
-Default behavior
+Stopping an epoch early
+----------------
+You can stop an epoch early by overriding :meth:`~pytorch_lightning.core.lightning.LightningModule.on_batch_start` to return -1 when some condition is met.
+
+If you do this repeatedly, for every epoch you had originally requested, then this will stop your entire run.
+
+Default Epoch End Callback Behavior
 ----------------
```
Comment (Borda) [inline docs/source/early_stopping.rst:11]:
```suggestion
Default Epoch End Callback Behavior
----------------------------------
```

File: docs/source/early_stopping.rst [hunk 4]
```diff
@@ -1,15 +1,21 @@
 Early stopping
 ==============
 
-Default behavior
+Stopping an epoch early
+----------------
+You can stop an epoch early by overriding :meth:`~pytorch_lightning.core.lightning.LightningModule.on_batch_start` to return -1 when some condition is met.
```
Comment (Borda) [inline docs/source/early_stopping.rst:6]:
```suggestion
Stopping an epoch early
-----------------------
You can stop an epoch early by overriding :meth:`~pytorch_lightning.core.lightning.LightningModule.on_batch_start` to return `-1` when some condition is met.
```

File: docs/source/early_stopping.rst [hunk 5]
```diff
@@ -1,15 +1,21 @@
...
-Default behavior
-----------------
+Stopping an epoch early
+-----------------------
+You can stop an epoch early by overriding :meth:`~pytorch_lightning.core.lightning.LightningModule.on_batch_start` to return `-1` when some condition is met.
+
+If you do this repeatedly, for every epoch you had originally requested, then this will stop your entire run.
+
+Default Epoch End Callback Behavior
+----------------------------------
```
Comment (Borda) [inline docs/source/early_stopping.rst:11]:
```suggestion
-----------------------------------
```

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -221,9 +221,6 @@ def training_step(self, batch, batch_idx, hiddens):
                         "hiddens": hiddens  # remember to detach() this
                     }
 
-            You can also return a -1 instead of a dict to stop the current loop. This is useful
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:224]: so this doesn't work? maybe we should add back this functionality if it got removed?

Comment (david-alexander-white) [reply pytorch_lightning/core/lightning.py:224]: I don't think it does? Let's trace it:

training_step is only called, I think, in training_forward, which will at best return it (training_loop.py 634)

then training_forward gets called by optimizer_closure (training_loop.py 525) 
```
                    with self.profiler.profile('model_forward'):
                        output = self.training_forward(
                            split_batch, batch_idx, opt_idx, self.hiddens)

                    closure_loss = output[0]
                    progress_bar_metrics = output[1]
                    log_metrics = output[2]
                    callback_metrics = output[3]
                    self.hiddens = output[4]
```

looks a lot to me like that will just blow up if you try to return -1. 

even if not, the only thing that comes back from optimizer_closure is the loss, no kind of status or anything.

Now, this all lives inside run_training_batch, which is the thing that needs to return a -1 as its first output. And it looks to me like it's hard-coded to return 0 except in a few cases where it's hard coded to return -1 (one of which I've added to the docs in this PR)

I could be wrong but i'd be surprised.

As to whether or not to add it back, it looks to me like that would be pretty hard, and the on_batch_start solution really isn't too bad

Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:224]: ok, and should your functionality also be added to batch_end? 
In any case I think a test would be good that this doesn't get lost in the future like the -1 in training_step :)

Comment (david-alexander-white) [reply pytorch_lightning/core/lightning.py:224]: great idea. i don't have the time to figure that out right now, maybe you can pick up the torch?

Comment (Borda) [reply pytorch_lightning/core/lightning.py:224]: @awaelchli do you have a test in mind?
